### ISSUE TITLE
minor fix to ROC

### ISSUE THREAD

**Issue by edublancas** [ploomber/sklearn-evaluation#121]
I noticed that the constructor in `ROC` takes `y_true` and `y_score` as arguments, but this should not be the case. The idea is that the constructor should only take the data required for plotting. In this case, that's `fpr` and `tpr` (and they should not have default values).

The `from_raw_data` should take `y_true` and `y_score`, compute `fpr` and `tpr` and call the constructor.

The confusion probably comes from the `ConfusionMatrix` implementation. I had to do it that way since I made a release with an old API that needed change when we added the `plot_a + plot_b` feature.

**Comment by yafimvo** [comment]
I pushed a fix.

I'm not sure about the multi class roc though. 
According to the previous logic (old api), we drew the multi class curve in 2 steps, first we drew the avg curve and then the other n curvers. For each curve we used `fpr` and `tpr` values. 

So, to draw a new multi class roc using `fpr` and `tpr` values, I added a new keyword argument `roc_rates_n_classes`, since the avg `fpr` and `tpr` are not enough to re-create the plot as before.

**Comment by edublancas** [comment]
ah. this answer the question I posted on the PR!

let me think about it for a bit. I'm unsure what's the best solution here!

**Comment by edublancas** [comment]
Ok, so I think the solution is for `ROC` only to have `fpr` and `tpr`, then process them internally.

In the simplest case, `fpr` and `fpr` are arrays (or lists), this should draw a single ROC curve.

In the second case, `fpr` and `fpr` will be a list of dictionaries:

```
fpr = [{"values": [...], "label": "some-name"}, ...]
```

This structure allows us to support multi-class, and the micro-average can have `"label": "micro-average"`


This will also allow us to support the third use case, which is comparing two binary classifiers:

```
fpr = [{"values": [...], "label": "first-model"}, {"values": [...], "label": "second-model"}]
```

In this case, there is no micro-average, since that's only applicable to multi-class.

thoughts @yafimvo ? I think this solves the problem

**Comment by edublancas** [comment]
I shared my feedback in the original issue. This will solve the problem. The only thing to keep in mind is to keep backward compatibility, so we should keep all arguments for now, and we'll deprecate in 0.10. Since the serialization format will also change, we need to support the old format as well.

**Comment by yafimvo** [comment]
> In the second case, `fpr` and `fpr` will be a list of dictionaries:
> 
> ```
> fpr = [{"values": [...], "label": "some-name"}, ...]
> ```

Each `fpr` and `tpr` holds a list of values and a label? In this case, they will hold duplicated labels which will be hard to maintain, no?

For the 2nd case, instead of using a list of dictionaries, can we hold a list of arrays?
then, we can use it something like this?

```python
roc.from_raw_data([[0.0, 0.2], [0.4, 0.4]], [[0.0, 0.2], [0.4, 0.4]], labels=['first-model', 'second-model'])
```

for the 1st case, it will be almost the same (with default label)

```python
roc.from_raw_data([0.0, 0.2], [0.4, 0.4])
```
@edublancas

**Comment by neelasha23** [comment]
> Each `fpr` and `tpr` holds a list of values and a label? In this case, they will hold duplicated labels which will be hard to maintain, no?

I guess labels would be corresponding class name?

**Comment by edublancas** [comment]
I see your point @yafimvo about duplicated labels. I like your idea of a list of arrays/lists

```python
roc.from_raw_data([[0.0, 0.2], [0.4, 0.4]], [[0.0, 0.2], [0.4, 0.4]], labels=['first-model', 'second-model'])
```

> I guess labels would be corresponding class name?

yes. in the multi-class case labels would be class name. But if we are comparing binary classifiers, then the labels can be the model name (e.g., xgboost, random forest, etc).

**Comment by yafimvo** [comment]
@edublancas so 2 arrays for `fpr` and `tpr` with optional variable: `labels`?

for the 1st case (single curve roc), do we want to use arrays of arrays (it is more consistent) or just an array of values (it is simpler)?

**Comment by edublancas** [comment]
> so 2 arrays for fpr and tpr with optional variable: labels?

yes

> for the 1st case (single curve roc), do we want to use arrays of arrays (it is more consistent) or just an array of values (it is simpler)?

i think it should work in both cases and have the same result, we should make things simple for users.

**Comment by idomic** [comment]
+1 on the second point.
